packages/lit-html/src/test/directives/async-append_test.ts

```diff
@@ -138,18 +142,144 @@ suite('asyncAppend', () => {
+        '<p>static</p>'
+      );
+      part.setConnected(false);
+      render(component(asyncAppend(iterable)), container);
+      await iterable.push('1');
+      assert.equal(
+        stripExpressionMarkers(container.innerHTML),
+        '<p>static</p>'
+      );
+      part.setConnected(true);
+      await nextFrame();
+      assert.equal(stripExpressionMarkers(container.innerHTML), '<p>1</p>');
+      await iterable.push('2');
+      assert.equal(stripExpressionMarkers(container.innerHTML), '<p>12</p>');
+    });
+
+    test('does not render when resolved and changed while disconnected', async () => {
+      const component = (value: any) => html`<p>${value}</p>`;
+      const part = render(component('staticA'), container);
+      assert.equal(
+        stripExpressionMarkers(container.innerHTML),
+        '<p>staticA</p>'
+      );
+      part.setConnected(false);
+      render(component(asyncAppend(iterable)), container);
+      await iterable.push('1');
+      assert.equal(
+        stripExpressionMarkers(container.innerHTML),
+        '<p>staticA</p>'
+      );
+      render(component('staticB'), container);
+      assert.equal(
+        stripExpressionMarkers(container.innerHTML),
+        '<p>staticB</p>'
+      );
+      part.setConnected(true);
+      await nextFrame();
+      assert.equal(
+        stripExpressionMarkers(container.innerHTML),
+        '<p>staticB</p>'
+      );
+      await iterable.push('2');
+      assert.equal(
+        stripExpressionMarkers(container.innerHTML),
+        '<p>staticB</p>'
+      );
+    });
+
+    test('the same promise can be rendered into two asyncAppend instances', async () => {
+      const component = (iterable: AsyncIterable<unknown>) =>
+        html`<p>${asyncAppend(iterable)}</p><p>${asyncAppend(iterable)}</p>`;
+      render(component(iterable), container);
+      assert.equal(
+        stripExpressionMarkers(container.innerHTML),
+        '<p></p><p></p>'
+      );
+      await iterable.push('1');
+      assert.equal(
+        stripExpressionMarkers(container.innerHTML),
+        '<p>1</p><p>1</p>'
+      );
+      await iterable.push('2');
+      assert.equal(
+        stripExpressionMarkers(container.innerHTML),
+        '<p>12</p><p>12</p>'
+      );
+    });
+  });
+
+  memorySuite('memory leak tests', () => {
+    test('tree with asyncAppend cleared while iterables are pending', async () => {
+      const template = (v: unknown) => html`<div>${v}</div>`;
+      // Make a big array set on an expando to exaggerate any leaked DOM
+      const big = () => new Array(10000).fill(0);
+      // Hold onto the iterables to prevent them from being gc'ed
+      const iterables: Array<TestAsyncIterable<string>> = [];
+      window.gc();
+      const heap = performance.memory.usedJSHeapSize;
+      for (let i = 0; i < 1000; i++) {
+        // Iterable passed to asyncAppend that will never yield
+        const iterable = new TestAsyncIterable<string>();
```

Comment:
Why do we create a new iterable per loop? Can we test with one iterable that never receives a value?

Reply:
The iterable is passed into the `until` and then the entire part is cleared in the same loop. Each loop makes a new `span (with expando) <-- part <-- directive instance (with iterable) <-- awaited closure` which will leak if the closure strongly references the directive instance.